### PR TITLE
Migrate `src/api` to TypeScript

### PR DESCRIPTION
**Pull request by ChrisAntaki**

### CODE REVIEW

File: src/runtime/basic-runtime-test.js [hunk 1]
```diff
@@ -155,22 +163,6 @@ describes.realWin('installBasicRuntime', (env) => {
     await getBasicRuntime().whenReady();
     expect(progress).to.equal('123');
   });
-
-  it('should implement all APIs', async () => {
```
Comment (ChrisAntaki) [inline src/runtime/basic-runtime-test.js:159]: Note: This test was essentially a copy of the "should implement BasicSubscriptions interface" test

File: src/runtime/runtime-test.js [hunk 2]
```diff
@@ -184,22 +228,6 @@ describes.realWin('installRuntime', (env) => {
     await getRuntime().whenReady();
     expect(progress).to.equal('123');
   });
-
-  it('implements all APIs', async () => {
```
Comment (ChrisAntaki) [inline src/runtime/runtime-test.js:188]: Note: This test was essentially a copy of the "should implement Subscriptions interface" test

File: src/runtime/basic-runtime-test.js [hunk 3]
```diff
@@ -105,16 +101,24 @@ describes.realWin('installBasicRuntime', (env) => {
     expect(getBasicRuntime()).to.equal(runtime1);
   });
 
+  // TODO: Delete this test after the TypeScript migration.
```
Comment (fikus) [inline src/runtime/basic-runtime-test.js:104]: Please reference a bug ID in TODOs.

File: src/runtime/runtime-test.js [hunk 4]
```diff
@@ -132,16 +131,60 @@ describes.realWin('installRuntime', (env) => {
     expect(getRuntime()).to.equal(runtime1);
   });
 
+  // TODO: Delete this test after the TypeScript migration.
```
Comment (fikus) [inline src/runtime/runtime-test.js:134]: Same here, reference a bug ID.

File: src/api/subscriptions.ts [hunk 5]
```diff
@@ -0,0 +1,572 @@
...
+
+  /**
+   * Starts the Offers flow.
+   */
+  showOffers(options?: OffersRequest): Promise<void>;
+
+  /**
+   * Starts the Offers flow for a subscription update.
+   * @param {!OffersRequest=} options
+   * @return {?}
```
Comment (fikus) [inline src/api/subscriptions.ts:81]: Remove or update.

File: src/api/subscriptions.ts [hunk 6]
```diff
@@ -0,0 +1,572 @@
...
+  showBestAudienceAction(): void;
+
+  /**
+   * Sets the publisherProvidedId.
+   */
+  setPublisherProvidedId(publisherProvidedId: string): Promise<void> | void;
+}
+
+export enum ShowcaseEvent {
+  // Events indicating content could potentially be unlocked
```
Comment (fikus) [inline src/api/subscriptions.ts:304]: nit: Change these descriptions to JSDoc.

File: src/api/subscriptions.ts [hunk 7]
```diff
@@ -0,0 +1,572 @@
...
+ *   The "auto" value either uses a redirect or a popup flow depending on
+ *   what's possible on a specific environment. Defaults to "auto".
+ * - enableSwgAnalytics - if set to true then events logged by the publisher's
+ *   client will be sent to Google's SwG analytics service.  This information is
+ *   used to compare the effectiveness of Google's buy-flow events to those
+ *   generated by the publisher's client code.  This includes events sent to
+ *   both PropensityApi and LoggerApi.
+ * - enablePropensity - If true events from the logger api are sent to the
+ *   propensity server.  Note events from the legacy propensity endpoint are
+ *   always sent.
```
Comment (fikus) [inline src/api/subscriptions.ts:361]: Move to JSDoc. Same for the other examples in this file.

File: src/api/subscriptions.ts [hunk 8]
```diff
@@ -0,0 +1,578 @@
...
+/**
+ * Properties:
+ * - skus - a list of SKUs to return from the defined or default list. The
+ *   order is preserved. Required if oldSku is specified (to indicate which
+ *   SKUs the user can upgrade or downgrade to).
+ * - list - a predefined list of SKUs. Use of this property is uncommon.
+ *   Possible values are "default" and "amp". Default is "default".
+ * - isClosable - a boolean value to determine whether the view is closable.
+ * - oldSku - Optional. The SKU to replace. For example, if a user wants to
+ *   upgrade or downgrade their current subscription.
```
Comment (fikus) [inline src/api/subscriptions.ts:498]: Move to JSDoc.

File: src/api/subscriptions.ts [hunk 9]
```diff
@@ -0,0 +1,578 @@
...
+ */
+export type SaveSubscriptionRequestCallback = () =>
+  | Promise<SaveSubscriptionRequest>
+  | SaveSubscriptionRequest;
+
+/**
+ * Properties:
+ * - lang: Sets the button SVG and title. Default is "en".
+ * - theme: "light" or "dark". Default is "light".
+ * - disable: whether to grey out the button.
```
Comment (fikus) [inline src/api/subscriptions.ts:533]: JSDoc.

File: src/api/subscriptions.ts [hunk 10]
```diff
@@ -0,0 +1,578 @@
...
+  theme?: string;
+  lang?: string;
+  disable?: boolean;
+}
+
+/**
+ * Properties:
+ * - lang: Sets the button SVG and title. Default is "en".
+ * - theme: "light" or "dark". Default is "light".
+ * - messageTextColor: Overrides theme color for message text. (ex: "#09f")
```
Comment (fikus) [inline src/api/subscriptions.ts:545]: JSDoc.

File: src/api/subscriptions.ts [hunk 11]
```diff
@@ -0,0 +1,578 @@
...
+/**
+ * Properties:
+ * - sku: Required. Sku to add to the user's subscriptions.
+ * - oldSku: Optional. This is if you want to replace one sku with another. For
+ *  example, if a user wants to upgrade or downgrade their current subscription.
+ * - prorationMode: Optional. When replacing a subscription you can decide on a
+ *  specific proration mode to charge the user.
+ *  The default is IMMEDIATE_WITH_TIME_PRORATION.
+ * - oneTime: Optional. When a user chooses a contribution, they have the option
+ *  to make it non-recurring.
```
Comment (fikus) [inline src/api/subscriptions.ts:562]: JSDoc.

**Review by fikus — APPROVED** [review]

File: src/api/basic-subscriptions.ts [hunk 12]
```diff
@@ -113,26 +103,20 @@ export const AutoPromptType = {
...
- *   allowScroll: (boolean|undefined),
- * }}
  */
-export let ClientOptions;
+export interface ClientOptions {
+  disableButton: boolean | undefined;
+  lang: string | undefined;
+  forceLangInIframes: boolean | undefined;
+  theme: ClientTheme | undefined;
+  allowScroll: boolean | undefined;
```
[the remark this answers is not in the document]
Comment (ChrisAntaki) [reply src/api/basic-subscriptions.ts:112]: Good call

**Pull request merged**